### ISSUE TITLE
How to make LeelaZero good at handicap games and good when ahead or behind

### ISSUE THREAD

**Issue by Nazgand** [leela-zero/leela-zero#1515]
A problem with AlphaZero and LeelaZero is that they play bad moves when they think they will win anyway or lose anyway. This is a problem in handicap games such as [Leela Zero Match (Game 5/8) - Haylee's Live Go 161](https://www.youtube.com/watch?v=2TDs7hRw_SI). Another problem is komi is unchangeable. Fair komi is a number that would produce ties under perfect play.

I want a Go playing AI that maximizes the number of points it obtains on the board. The following implementations could solve this problem.

-------
Option 1:
Have almost 361 neural nets. Let PFMCP(n) be the Probability of ability to Force a Minimum of n Chinese Points.
PFMCP(0) should always output 1, and can thus be ignored.
The minimum number of points one could have greater than 0 is 3 points as the smallest seki, where the opponents gets 357 points, so only PFMCP(3) through PFMCP(358) need to be trained.

When choosing a move:
There is some cutoff, c, e.g. 0.99, but probably lower when beginning training.
Find the minimum n such that for all m>n, PFMCP(m)<c, Select the set of moves for which result in PFMCP(n)>c, then select the move which maximizes PFMCP(n+1).

Possible shortcut seed:
If no neutral territory exists, komi is 7.5, and white wins, then white has at least 177 points on the board. Thus seeding PFMCP(177) with the current LeelaZero network seems reasonable. Discussion requested on pros and cons.

All other details would be implemented the same way as LeelaZero, and thus is probably easier than Option 2.

-------
Option 2:
A single neural net could output, for each move, the number of points it thinks it can forcefully obtain. This option needs more careful thought to train than the first option, yet is probably more computationally efficient. I would appreciate discussion of the possible implementations of the neural net architecture and details of the training process.

-------
Both options:
During training:
In addition to starting with an empty board, some training games would start in the standard handicap positions, and others would start with a randomly generated board position.
A possible nice feature would be to allow people to choose the starting positions with an AutoGTP option, allowing anyone to train LeelaZero in position in which LeelaZero has been observed to make a suboptimal move.

When starting to train, random positions with a large number of stones, e.g. >300, would be good because the games would be short, thus speeding up the training of endgame ability. 

-------
Would anyone else be interested in a LeelaZero fork like this? I suspect more people will be interested after LeelaZero is more fully trained, which may take a few more years. Planning the next version now is nice.
Some might like the way that LeelaZero gambles for the win when LeelaZero thinks it is behind, even though 7.5 is an arbitrary and obviously unfair komi. The proposed modified LeelaZero would not know or care whether it is winning and would just strive to maximize points.

**Comment by alreadydone** [comment]
I am copying my proposal from https://github.com/gcp/leela-zero/issues/1470#issuecomment-392452882 a week ago; no one objected so far, and I still think it's a feasible plan. If this has to happen in a fork I am ready to use Big Red II to support it. I should be able to modify the code of the leelaz engine 
to let it continue to play after resignation using the score head, but I probably won't be able to the training.

------------------------------
What about the following proposal? Since the network already knows how to count, to extract the knowledge of score from the network we just introduce a "score head" with identical architecture as the value head except for omitting the tanh, which still needs to be trained though. We can randomly initialize its weights or maybe copy from the value head, and for training data we can just continue to play resigned games, first using the policy only, and then using the score head as well (to maximize expected score) after it is sufficiently well trained. Since the resigned games have winrates below 10% or so, this hopefully doesn't mess up the policy in normal situations. It doesn't matter what policy suggests in post-resignation situations, and aiming towards higher score seems a good way to make the suggestions more reasonable; also maximizing the score seems to be a goal that will induce aggressive play in handicapped games.

We can use a small weighting for samples drawn from these post-resignation positions so as to minimize any potential adverse effects on normal training; depending on whether it helps normal playing strength or not, we can decide whether to raise or lower the weighting. Or we can freeze the other weights when training the score head to eliminate any effect.

During handicap playing, we can play to maximize some reward function calculated from the score head, and switch to base on winrate once winrate rises above resignation threshold (or higher, say 20%). For example, if the score head says you are n points behind you should prefer 50% winning by 0.5 points and 50% losing by 0.5+2n points to 100% losing by n points, and some reward function like tanh(x+(n-0.5)/2) should work. In addition to score estimation and handicap playing, the score head can also be used to play under any komi, and for perfect endgame.

**Comment by betterworld** [comment]
> Option 2:
A single neural net could output, for each move, the number of points it thinks it can forcefully obtain.

How would this include risk and luck etc? There will be positions where I can say I can win by 100 with 50% or win by 10 with 80%, and of course it always depends on the strength of my opponent, especially since we are discussing handicap games.

> During training:
In addition to starting with an empty board, some training games would start in the standard handicap positions

This is probably the best way to get a good handicap bot. I'd suggest that in these games the two players will use some sort of randomized imbalance, e.g. different number of playouts for black and white. Black should only be the weaker player in half of the games, otherwise the network will learn that Black is generally more likely to be a weaker player regardless of the number of handicap stones.

**Comment by Nazgand** [comment]
@alreadydone 

> During handicap playing, we can play to maximize some reward function calculated from the score head, and switch to base on winrate once winrate rises above resignation threshold (or higher, say 20%). 

This seems good, and I would argue to symmetrically use the value head rather than the winrate when the winrate is large e.g.>80%. This may be what you meant by `and for perfect endgame`. With a score head implemented, I could get the sort of bot I want with an option that maximizes score rather than gambling to win. If the option is not built in, I should be able to set komi to 400 points in my favour to get similar the same behaviour.

@betterworld 

> How would this include risk and luck etc? There will be positions where I can say I can win by 100 with 50% or win by 10 with 80%, and of course it always depends on the strength of my opponent, especially since we are discussing handicap games.

My vision was a bot that (plays under the assumption that the opponent is just as good as it) and never gambles points because it (doesn't know or care what komi is) and doesn't know or care what winning is, only aiming for the maximum forcible points. In practice, the maximum forcible points will increase whenever the weaker player makes mistakes, so purposely playing bad moves which raise the probability of a weaker player making even worse moves is likely unnecessary.

Avoiding gambling also simplifies through not needing to figure out how much to gamble based on the inferred strength of the opponent. Perhaps the weakness could be approximated by examining differences between winrates of perceived best moves and winrates of the opponent's chosen moves, or differences betweens a score head's evaluation of moves. A running calculation of the average and standard deviation of the perceived numbers of points lost per move could be a simple guideline.

**Comment by Ishinoshita** [comment]
"purposely playing bad moves which raise the probability of a weaker player making even worse moves is likely unnecessary.
In practice, at handicap,  even if white plays 'honestly', that is avoids too much overplay, he cannot just passively wait for black mistakes. He will often tenuki and let some of his stones, groups unsettled, floating. If he goes too far in a losing game, he will usually try to complicate the game, start fights, etc....  before resigning. He will usually favor complex lines of play, involving exchanges, to have a chance that black make missteps. Black under optimal moves is often

**Comment by alreadydone** [comment]
Discussion on overplay on reddit: https://www.reddit.com/r/baduk/comments/8oij57/leela_zero_match_game_68_haylees_live_go_162/

**Comment by willemhendriks** [comment]
For humans - white tries to force a complicated games - situations where black options are less obvious and prone to suboptimal moves.

I was thinking - If during handicap, Black's policy network is replaced with a significant weaker one - maybe White will explore those type of moves we want him to play more.

**Comment by sethtroisi** [comment]
closing older issues, no discussion for >6 months
